contracts/contracts/keeper/Keeper.sol

```diff
@@ -0,0 +1,194 @@
+    //  * the sender is the zero address.
+    //  */
+    // function preventExecution() internal view {
+    //   console.log('In preventExecution');
+    //   console.log(tx.origin);
+    //   require(tx.origin == address(0), "only for simulated backend");
+    // }
+
+    constructor(address payable _vaultCoreAddr, address payable _ousdAddr)
+        public
+    {
+        vaultCore = VaultCore(_vaultCoreAddr);
+        oUSD = OUSD(_ousdAddr);
+        nextAllocate = now + 1 days;
+    }
+
+    /*
+     * @notice method that is simulated by the keepers to see if any work actually
+     * needs to be performed. This method does does not actually need to be
+     * executable, and since it is only ever simulated it can consume lots of gas.
+     * @dev To ensure that it is never called, you may want to add the
+     * cannotExecute modifier from KeeperBase to your implementation of this
+     * method.
+     * @return success boolean to indicate whether the keeper should call
+     * performUpkeep or not.
+     * @return success bytes that the keeper should call performUpkeep with, if
+     * upkeep is needed.
+     */
+
+    function checkUpkeep(bytes calldata _data)
+        external
+        returns (
+            // view
+            //cannotExecute
+            bool success,
+            bytes memory dynamicData
+        )
+    {
+        bool rebase = shouldRebase();
+        bool allocate = shouldAllocate();
+        bool isSuccessful = rebase || allocate;
+
+        ActionInfo memory rebaseInfo = ActionInfo({
+            actionType: REBASE,
+            run: rebase,
+            param: "foo"
+        });
+        ActionInfo memory allocateInfo = ActionInfo({
+            actionType: ALLOCATE,
+            run: allocate,
+            param: "foo"
+        });
+        KeepInfo memory keepInfo = KeepInfo({
+            rebaseInfo: rebaseInfo,
+            allocateInfo: allocateInfo
+        });
+
+        bytes memory data = abi.encode(keepInfo);
+        return (isSuccessful, data);
+    }
+
+    function performUpkeep(bytes calldata dynamicData) external {
+        KeepInfo memory keepInfo = abi.decode(dynamicData, (KeepInfo));
+
+        require(
+            keepInfo.rebaseInfo.run || keepInfo.allocateInfo.run,
+            "No keeper actions are enabled"
+        );
+        require(
+            shouldRebase() || shouldAllocate(),
+            "No keeper actions are callable"
+        );
+
+        if (keepInfo.rebaseInfo.run) {
+            vaultCore.rebase();
+        }
+
+        if (keepInfo.allocateInfo.run) {
+            vaultCore.allocate();
+            nextAllocate = now + 1 days;
+        }
+    }
+
+    function shouldRebase() internal view returns (bool) {
+        uint256 vaultValue = vaultCore.totalValue();
+        uint256 ousdSupply = oUSD.totalSupply();
+
+        if (ousdSupply == 0) {
+            return false;
+        }
+
+        bool rebase = (vaultValue + rebaseThreshold) > ousdSupply;
```

Comment:
We generally try to use SafeMath everywhere, even if not really required :)

Reply:
done- thanks